posthog/async_migrations/migrations/0004_replicated_schema.py

```diff
@@ -0,0 +1,312 @@
+            ),
+            AsyncMigrationOperation.simple_op(sql="SYSTEM START MERGES", rollback="SYSTEM STOP MERGES",),
+        ]
+
+    def replicated_table_operations(self, table: TableMigrationData):
+        yield AsyncMigrationOperation.simple_op(
+            sql=f"""
+            CREATE TABLE {table.tmp_table_name} AS {table.name}
+            ENGINE = {self.get_new_engine(table)}
+            """,
+            rollback=f"DROP TABLE IF EXISTS {table.tmp_table_name}",
+        )
+
+        if table.materialized_view_name is not None:
+            yield AsyncMigrationOperation.simple_op(
+                sql=f"DROP TABLE IF EXISTS {table.materialized_view_name}", rollback=table.create_materialized_view,
+            )
+
+        yield AsyncMigrationOperation(
+            fn=lambda _: self.move_partitions(table.name, table.tmp_table_name),
+            rollback_fn=lambda _: self.move_partitions(table.tmp_table_name, table.name),
+        )
+        yield AsyncMigrationOperation.simple_op(
+            sql=f"RENAME TABLE {table.name} TO {table.backup_table_name}, {table.tmp_table_name} TO {table.renamed_table_name}",
+            rollback=f"RENAME TABLE {table.backup_table_name} TO {table.name}, {table.renamed_table_name} TO {table.tmp_table_name}",
+        )
+
+        if table.materialized_view_name is not None:
+            yield AsyncMigrationOperation.simple_op(
+                sql=table.create_materialized_view, rollback=f"DROP TABLE IF EXISTS {table.materialized_view_name}",
+            )
+
+        # NOTE: Relies on IF NOT EXISTS on the query
+        if isinstance(table, ShardedTableMigrationData):
+            for create_table_query in table.extra_tables:
+                yield AsyncMigrationOperation.simple_op(sql=create_table_query)
+
+    def get_current_engine(self, table_name: str):
+        return sync_execute(
+            "SELECT engine_full FROM system.tables WHERE database = %(database)s AND name = %(name)s",
+            {"database": settings.CLICKHOUSE_DATABASE, "name": table_name},
+        )[0][0]
+
+    def get_new_engine(self, table: TableMigrationData):
+        """
+        Returns new table engine statement for the table.
+
+        Note that the engine statement also includes PARTITION BY, ORDER BY, SAMPLE BY and SETTINGS,
+        so we use the current table as a base for that and only replace the
+        """
+        current_engine = self.get_current_engine(table.name)
+
+        if "Replicated" in current_engine or "Distributed" in current_engine:
+            raise ValueError(
+                f"""
+                Table engine of incorrect type, cannot be replicated or distributed.
+
+                table={table.name}, current_engine={current_engine}
+            """
+            )
+
+        # Remove the current engine from the string
+        return re.sub(r".*MergeTree\(\w+\)", str(table.new_table_engine), current_engine)
+
+    def move_partitions(self, from_table: str, to_table: str):
```

Comment:
sth to consider: if this fails we won't get a very nice error for the user probably, which is why I added the try catch and log sql too into execute_clickhouse_op etc.

Reply:
I highlighted this in my feedback - execute_op_* is not a good API and wouldn't work here without significant refactoring and also getting rid of it requires also changing other migrations.

Reply:
you don't have to use `execute_op` but it could be nice to check that we give users/us enough info to debug why this step failed